### ISSUE TITLE
Mimic "Google Scholar PDF Reader" features

### ISSUE THREAD

**Issue by 7vq** [mozilla/pdf.js#17921]
Chrome has a new ["Google Scholar PDF Reader" extension](https://chromewebstore.google.com/detail/dahenjhkoodjbpjheillcadbppiidmhp) for academic articles. The extension allows clicking in-text citations to show, in a popup: the cited article's title, full author list, and article summary. That avoids losing your reading place every time you want to see what paper a citation corresponds to. It also avoids having to manually copy, paste, search, and find each citation online to read the abstract. The extension also automatically adds a linked table of contents for the current PDF, even for PDFs without a built-in linked table of contents.

To convey how much people like these features: the extension was only released a few months ago and already has 100k users.

Could these features be added to pdf.js? (It sounds to me like Firefox [wouldn't allow](https://bugzilla.mozilla.org/show_bug.cgi?id=1454760) these features to be implemented as a Firefox extension?)

**Comment by sylvestre** [comment]
FYI, this Google extension is based on pdf.js

**Comment by jacklanda** [comment]
Comment for better reach.

**Comment by JasonSome** [comment]
Any update/news on this pending feature?

**Comment by marco-c** [comment]
No news, and we have other things with higher priorities in our roadmap at the moment.

**Comment by Snuffleupagus** [comment]
> [This bookmarklet](https://github.com/belinghy/PDFRefPreview) gives a popup preview for internal links like figures and citations. The only issue is I haven't found a way to automate manually clicking/calling the bookmarklet on every PDF load.
>
> Maybe that repository's code could be integrated into pdf.js?

I don't think so, since that code can end up trigger rendering *a lot* when just moving the mouse around which may hurt overall performance in many cases.
This is made worse by the fact that there's no attempt, as far as I can tell, to actually cancel (and clean-up) properly once the preview is no longer needed.

**Comment by 7vq** [comment]
> I don't think so, since that code can end up trigger rendering _a lot_ when just moving the mouse around which may hurt overall performance in many cases. This is made worse by the fact that there's no attempt, as far as I can tell, to actually cancel (and clean-up) properly once the preview is no longer needed.

Ah, fair enough. (For what it's worth, I haven't noticed any issues myself in years of usage. But obviously other people will have different usages, plus I might just be overlooking issues.)

**Comment by salomepoulain** [comment]
Any updates?

**Comment by yuyuslab** [comment]
Is Google Scholar PDF reader extension an open source project??
I'm thinking of adding annotation features somehow...

**Comment by salcc** [comment]
I've created a modified version of the Google Scholar PDF Reader extension that adds annotation capabilities while maintaining the core citation preview features: https://github.com/salcc/Scholar-PDF-Reader-with-Annotations. The extension is open source - feel free to use it or contribute!

*Sorry for the slight off-topic reply, but I thought many people that come here just want these features (like me a few months ago)*

**Comment by yuyuslab** [comment]
Wow, I will take a look soon!!
Thank you a lot 😆

**Comment by yuyuslab** [comment]
I've just used the extension, and it works just fine!! thx for your work!
I might create PR for a new feature later.

**Comment by JasonSome** [comment]
Has this feature been prioritized in the roadmap yet? If so, where is it placed? The academic and industrial research community is somewhat hindered by this missing technical implementation, which might affect Firefox's usage patterns.

**Comment by marco-c** [comment]
@JasonSome no, it hasn't been prioritized yet. We'd happily accept contributions though.

**Comment by alex180500** [comment]
Please please implement this, it's so useful to copy citations directly from the pdf reader, I use it all the time!!! It's the only thing that's keeping my google chrome installed.

**Comment by salcc** [comment]
Alternatively, Firefox could support the `sandbox` content security policy in the manifest.json for web extensions which would make it possible for the Google Scholar PDF Reader extension to support Firefox. See https://bugzilla.mozilla.org/show_bug.cgi?id=1685123.

**Comment by francescotaioli** [comment]
Google Scholar's PDF Reader is very useful for reading papers. This would be a great feature."

**Comment by BuiQuocDung1991** [comment]
I am using both Firefox and Chrome simultaneously, but that Chrome extension is really great, offering many benefits for users, such as dark mode, citation, AI-generated outlines, and sections. I hope these features will be added to PDF.js because I think that Firefox’s dull

**Comment by alex180500** [comment]
Any news on this? is it so hard to implement?

**Comment by sylvestre** [comment]
If a company wants to fund this, please contact me: s@mozilla.com

For now, we are focusing on other features that we think are more important for our users.

**Comment by zmtomorrow** [comment]
Hi, do we have any ideas of how Google generate TOC, even for PDFs without a built-in linked table of contents? We could try to contribute if we have some clues of how they do that.

**Comment by Lr-2002** [comment]
could it be support later? it is really necessary for us !
